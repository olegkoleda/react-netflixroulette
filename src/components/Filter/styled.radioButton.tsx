import styled from "styled-components";

export const StyledRadioButton = styled.div<{ checked?: boolean }>`
  text-align: center;
  padding: 0.7rem;
  margin-right: 1.5rem;
  border-bottom: ${(props) =>
    props.checked ? `2px solid ${props.theme.color.red}` : ""};

  input[type="radio"] {
    padding: 0;
    overflow: hidden;
    clip: rect(0, 0, 0, 0);
    white-space: nowrap;
    clip-path: inset(50%);
    border: 0;
    position: absolute;
    width: 0.0625rem;
    height: 0.0625rem;

    &:focus {
      + label {
        outline: 0.125rem solid ${(props) => props.theme.color.white};
        outline-offset: 0.25rem;
      }
    }
  }

  label {
    cursor: pointer;
  }
`;
