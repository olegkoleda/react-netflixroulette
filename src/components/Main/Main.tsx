import React from "react";
import { StyledMain, StyledResults } from "./styled.main";
import { Composition, Box } from "atomic-layout";
import Filter from "../Filter";
import Sort from "../Sort";
import CardsList from "../CardsList";

const mockedData = {
  filters: [
    { filter: "All", checked: true },
    { filter: "Documentary", checked: false },
    { filter: "Comedy", checked: false },
    { filter: "Horror", checked: false },
  ],
  sort: ["Release date", "Rating", "Budget"],
};

const controlsAreasMd = `filter filter filter sort`;

const controlsAreasMobile = `
  filter
  sort
`;

const Main = () => (
  <Box as={StyledMain} paddingHorizontal={"1rem"} paddingHorizontalMd={"3rem"}>
    <Composition areas={controlsAreasMobile} areasMd={controlsAreasMd}>
      {(Areas) => (
        <>
          <Areas.Filter>
            <Filter filters={mockedData.filters} />
          </Areas.Filter>
          <Areas.Sort>
            <Sort sortValues={mockedData.sort} />
          </Areas.Sort>
        </>
      )}
    </Composition>
    <StyledResults>
      <span>39</span> movies found
    </StyledResults>
    <CardsList />
  </Box>
);

export default Main;
