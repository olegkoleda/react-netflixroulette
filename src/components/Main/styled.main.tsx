import styled from "styled-components";

export const StyledMain = styled.main`
  height: 100%;
  min-height: calc(50vh - 6.5rem);
  background: ${(props) => props.theme.color.black};
`;

export const StyledResults = styled.div`
  margin: 1.875rem 0;
  font-size: 1.125rem;

  & span {
    font-weight: 600;
  }
`;
