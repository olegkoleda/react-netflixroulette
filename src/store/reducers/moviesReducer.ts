import {
  LOAD_MOVIES_ERROR,
  LOAD_MOVIES_STARTED,
  LOAD_MOVIES_SUCCESS,
  SET_ACTIVE_MOVIE,
  LoadMovieActionType,
  ISetActiveMovie,
} from "../types/loadMovies";
import {
  ADD_MOVIE_ERROR,
  ADD_MOVIE_STARTED,
  ADD_MOVIE_SUCCESS,
  ADD_MOVIE_FINISHED,
  AddMovieActionType,
} from "../types/addMovie";
import {
  DELETE_MOVIE_ERROR,
  DELETE_MOVIE_STARTED,
  DELETE_MOVIE_SUCCESS,
  DeleteMovieActionType,
} from "../types/deleteMovie";
import { IMovie } from "../../interfaces/IMovie";
import {
  UPDATE_MOVIE_ERROR,
  UPDATE_MOVIE_STARTED,
  UPDATE_MOVIE_SUCCESS,
  UpdateMovieActionType,
  UPDATE_MOVIE_FINISHED,
} from "../types/updateMovie";

export interface IMoviesState {
  list: IMovie[];
  loading: boolean;
  isMovieUpdated: boolean;
  error: Error | null;
  activeMovie: number | null;
  isMovieAdded: boolean;
}

const initialState: IMoviesState = {
  list: [],
  loading: false,
  error: null,
  activeMovie: null,
  isMovieUpdated: false,
  isMovieAdded: false,
};

export default function moviesReducer(
  state = initialState,
  action:
    | LoadMovieActionType
    | ISetActiveMovie
    | AddMovieActionType
    | DeleteMovieActionType
    | UpdateMovieActionType
) {
  switch (action.type) {
    case LOAD_MOVIES_STARTED:
    case ADD_MOVIE_STARTED:
    case DELETE_MOVIE_STARTED:
    case UPDATE_MOVIE_STARTED:
      return {
        ...state,
        loading: true,
      };
    case LOAD_MOVIES_SUCCESS:
      return {
        ...state,
        loading: false,
        list: action.payload,
      };
    case LOAD_MOVIES_ERROR:
    case ADD_MOVIE_ERROR:
    case DELETE_MOVIE_ERROR:
    case UPDATE_MOVIE_ERROR:
      return {
        ...state,
        error: action.error,
      };
    case SET_ACTIVE_MOVIE:
      return {
        ...state,
        activeMovie: action.payload,
      };
    case UPDATE_MOVIE_FINISHED:
      return {
        ...state,
        isMovieUpdated: false,
      };
    case ADD_MOVIE_FINISHED:
      return {
        ...state,
        isMovieAdded: false,
      };
    case ADD_MOVIE_SUCCESS:
      return {
        ...state,
        loading: false,
        isMovieAdded: true,
        list: [action.payload, ...state.list],
      };
    case DELETE_MOVIE_SUCCESS:
      return {
        ...state,
        loading: false,
        list: state.list.filter(({ id }) => id !== action.payload),
      };
    case UPDATE_MOVIE_SUCCESS:
      return {
        ...state,
        loading: false,
        isMovieUpdated: true,
        list: state.list.map((movie) =>
          movie.id === action.payload.id ? action.payload : movie
        ),
      };
    default:
      return state;
  }
}
