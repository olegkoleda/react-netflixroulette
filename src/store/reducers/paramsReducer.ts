import { FilterMovieActionType, SET_MOVIE_FILTER } from "../types/filterMovie";

export interface IParamsState {
  filter: string;
  sortBy: string;
}

const initialState: IParamsState = {
  filter: "",
  sortBy: "release_date",
};

export default function paramsReducer(
  state = initialState,
  action: FilterMovieActionType
) {
  switch (action.type) {
    case SET_MOVIE_FILTER:
      return {
        ...state,
        filter: action.payload,
      };

    default:
      return state;
  }
}
