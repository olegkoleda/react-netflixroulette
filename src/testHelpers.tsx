import React from "react";
import renderer from "react-test-renderer";
import { theme } from "./styles/Theme";
import { ThemeProvider } from "styled-components";

export function renderWithTheme(component: React.ReactNode) {
  return renderer.create(
    <ThemeProvider theme={theme}>{component}</ThemeProvider>
  );
}
