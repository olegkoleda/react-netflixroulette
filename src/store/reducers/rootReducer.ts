import { combineReducers } from "redux";
import { IParamsState } from "./paramsReducer";
import { IMoviesState } from "./moviesReducer";
import movies, { initialState as initialMovieState } from "./moviesReducer";
import params, { initialState as initialParamsState } from "./paramsReducer";

export default combineReducers({
  movies,
  params,
});

export interface IAppState {
  movies: IMoviesState;
  params: IParamsState;
}

export const initialAppState: IAppState = {
  movies: initialMovieState,
  params: initialParamsState,
}
