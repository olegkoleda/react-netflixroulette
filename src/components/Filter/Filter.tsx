import React, { ChangeEvent, useCallback, useState } from "react";
import { StyledRadioButton } from "./styled.radioButton";
import { Box } from "atomic-layout";
import { IAppState } from "../../store/reducers/rootReducer";
import { connect } from "react-redux";
import { filterMovie } from "../../store/actions/filterMovie";
import { loadMovies } from "../../store/actions/loadMovies";

interface IFilterType {
  value: string;
  label: string;
}

interface IFilterProps {
  filters: Array<IFilterType>;
  activeFilter: string;
  setFilter: Function;
  loadMoviesProp: Function;
}

const Filter: React.FC<IFilterProps> = ({
  filters,
  activeFilter,
  setFilter,
  loadMoviesProp,
}) => {
  const changeValue = useCallback((e: ChangeEvent<HTMLInputElement>) => {
    setFilter(e.target.value);
    loadMoviesProp();
  }, []);

  return (
    <Box flex onChange={changeValue}>
      {filters.map(({ label, value }: IFilterType) => {
        const checked = value === activeFilter;
        const filterId = `filter_${label}`;
        return (
          <StyledRadioButton checked={checked} key={filterId}>
            <input
              defaultChecked={checked}
              type="radio"
              name="filter"
              value={value}
              id={filterId}
            />
            <label htmlFor={filterId}>{label.toUpperCase()}</label>
          </StyledRadioButton>
        );
      })}
    </Box>
  );
};

const mapState = (state: IAppState) => ({
  activeFilter: state.params.filter,
});

const mapDispatch = {
  setFilter: (filter: string) => filterMovie(filter),
  loadMoviesProp: () => loadMovies(),
};

export default connect(mapState, mapDispatch)(Filter);
