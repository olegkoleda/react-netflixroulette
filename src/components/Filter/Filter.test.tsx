import React from "react";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import { Provider } from "react-redux";
import Filter from "./Filter";
import { renderWithTheme } from "../../testHelpers";
import { initialAppState } from "../../store/reducers/rootReducer";
import { mockedData } from "../Main/Main";

const middleWares = [thunk];
const mockStore = configureMockStore(middleWares);
const store = mockStore(initialAppState);

test("Filters component matches the Snapshot", () => {
  const component = renderWithTheme(
    <Provider store={store}>
        <Filter filters={mockedData.filters} />
    </Provider>
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
