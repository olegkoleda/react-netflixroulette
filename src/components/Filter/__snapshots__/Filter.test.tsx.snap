// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters component matches the Snapshot 1`] = `
<div
  className="sc-AxjAm fRMFrE"
  onChange={[Function]}
>
  <div
    checked={true}
    className="styledradioButton__StyledRadioButton-sc-1athmhv-0 bdPXCN"
  >
    <input
      defaultChecked={true}
      id="filter_All"
      name="filter"
      type="radio"
      value=""
    />
    <label
      htmlFor="filter_All"
    >
      ALL
    </label>
  </div>
  <div
    checked={false}
    className="styledradioButton__StyledRadioButton-sc-1athmhv-0 bNqTgl"
  >
    <input
      defaultChecked={false}
      id="filter_Documentary"
      name="filter"
      type="radio"
      value="Documentary"
    />
    <label
      htmlFor="filter_Documentary"
    >
      DOCUMENTARY
    </label>
  </div>
  <div
    checked={false}
    className="styledradioButton__StyledRadioButton-sc-1athmhv-0 bNqTgl"
  >
    <input
      defaultChecked={false}
      id="filter_Comedy"
      name="filter"
      type="radio"
      value="Comedy"
    />
    <label
      htmlFor="filter_Comedy"
    >
      COMEDY
    </label>
  </div>
  <div
    checked={false}
    className="styledradioButton__StyledRadioButton-sc-1athmhv-0 bNqTgl"
  >
    <input
      defaultChecked={false}
      id="filter_Horror"
      name="filter"
      type="radio"
      value="Horror"
    />
    <label
      htmlFor="filter_Horror"
    >
      HORROR
    </label>
  </div>
</div>
`;
